staging/src/k8s.io/mount-utils/mount_linux_test.go

```diff
@@ -715,23 +699,104 @@ func TestFormat(t *testing.T) {
 			}
 			mounter := NewSafeFormatAndMount(nil, exec, WithMaxConcurrentFormat(tc.max, tc.timeout))
 
-			var wg sync.WaitGroup
-			for i := 0; i < formatCount; i++ {
-				wg.Add(1)
+			// we run max+1 goroutines and block the command execution
+			// only max goroutine should be running and the additional one should wait
+			// for one to be released
+			for i := 0; i < tc.max+1; i++ {
 				go func() {
-					defer wg.Done()
 					mounter.format(fstype, nil)
 				}()
 			}
-			wg.Wait()
 
-			if maxConcurrent != tc.wantConcurrent {
-				t.Errorf("SafeFormatAndMount.format() got concurrency: %d, want: %d", maxConcurrent, tc.wantConcurrent)
+			// wait for all goorutines to be scheduled
+			time.Sleep(100 * time.Millisecond)
+
+			mu.Lock()
+			if concurrent != tc.max {
+				t.Errorf("SafeFormatAndMount.format() got concurrency: %d, want: %d", concurrent, tc.max)
+			}
+			mu.Unlock()
+
+			// signal the commands to finish the goroutines, this will allow the command
+			// that is pending to be executed
+			for i := 0; i < tc.max; i++ {
+				witness <- struct{}{}
 			}
+
+			// wait for all goroutines to acquire the lock and decrement the counter
+			time.Sleep(100 * time.Millisecond)
+
+			mu.Lock()
+			if concurrent != 1 {
+				t.Errorf("SafeFormatAndMount.format() got concurrency: %d, want: 1", concurrent)
+			}
+			mu.Unlock()
+
+			// signal the pending command to finish, no more command should be running
+			close(witness)
+
+			// wait a few for the last goroutine to acquire the lock and decrements the counter down to zero
+			time.Sleep(10 * time.Millisecond)
+
+			mu.Lock()
+			if concurrent != 0 {
+				t.Errorf("SafeFormatAndMount.format() got concurrency: %d, want: 0", concurrent)
+			}
+			mu.Unlock()
 		})
 	}
 }
 
+func TestFormatTimeout(t *testing.T) {
+	const (
+		formatCount    = 5
+		fstype         = "ext4"
+		output         = "complete"
+		maxConcurrency = 4
+		timeout        = 200 * time.Millisecond
+	)
+
+	var concurrent int
+	var mu sync.Mutex
+	witness := make(chan struct{})
+
+	exec := &testexec.FakeExec{}
+	for i := 0; i < formatCount; i++ {
+		exec.CommandScript = append(exec.CommandScript, makeFakeCommandAction(output, nil, func() {
+			mu.Lock()
+			concurrent++
+			mu.Unlock()
+			<-witness
+		}))
+	}
+	mounter := NewSafeFormatAndMount(nil, exec, WithMaxConcurrentFormat(maxConcurrency, timeout))
+
+	for i := 0; i < maxConcurrency+1; i++ {
+		go func() {
+			mounter.format(fstype, nil)
+		}()
+	}
+
+	// wait for all goroutines to be scheduled
+	time.Sleep(100 * time.Millisecond)
```

Comment:
you have to wait here a bit more than the timeout specified and check that `concurrent` is equal to `maxConcurrency+1` , add also a comment with a TODO referencing or adding my comment about using the ratelimiter and the os.exec with context

Reply:
once you check that you just close the channel and verify that concurrency is 0

Reply:
@aojea Where I should I put the TODO above the function itself ? or you prefer in the test file ?
below the `TestFormatTimeout` modified according your last review:
```go
func TestFormatTimeout(t *testing.T) {
	const (
		formatCount    = 5
		fstype         = "ext4"
		output         = "complete"
		maxConcurrency = 4
		timeout        = 200 * time.Millisecond
	)

	var concurrent int
	var mu sync.Mutex
	witness := make(chan struct{})

	exec := &testexec.FakeExec{}
	for i := 0; i < formatCount; i++ {
		exec.CommandScript = append(exec.CommandScript, makeFakeCommandAction(output, nil, func() {
			mu.Lock()
			concurrent++
			mu.Unlock()

			<-witness

			mu.Lock()
			concurrent--
			mu.Unlock()
		}))
	}
	mounter := NewSafeFormatAndMount(nil, exec, WithMaxConcurrentFormat(maxConcurrency, timeout))

	for i := 0; i < maxConcurrency+1; i++ {
		go func() {
			mounter.format(fstype, nil)
		}()
	}

	// wait a bit more than the configured timeout
	time.Sleep(timeout + 100*time.Millisecond)

	mu.Lock()
	if concurrent != maxConcurrency+1 {
		t.Errorf("SafeFormatAndMount.format() got concurrency: %d, want: %d", concurrent, maxConcurrency+1)
	}
	mu.Unlock()

	// signal the pending commands to finish
	close(witness)
	// wait for all goroutines to acquire the lock and decrement the counter
	time.Sleep(100 * time.Millisecond)

	mu.Lock()
	if concurrent != 0 {
		t.Errorf("SafeFormatAndMount.format() got concurrency: %d, want: 0", concurrent)
	}
	mu.Unlock()
}
```

Reply:
test file sounds good

Reply:
done in 1c45bacfb0e